### DISCUSSION TITLE
Contact Distance Parameter

### DISCUSSION THREAD

**Discussion by vtnick** [idaholab/moose#25526] [Q&A Modules: Solid mechanics]
I am working a problem with thermal expansion between two bodies with a small initial gap (1E-4). Based on hand calculations and other FE solvers the two bodies will not come into contact but have a very small gap (1E-5), but with the MOOSE simulation the bodies do in fact come in to contact and at a lower heat input. Is there a parameter I am missing in the contact action block that controls the nodal distance when checking for contact?

**Comment by GiudGiud** [comment]
Hello

Is your result in MOOSE converged?
In terms of:
- mesh density
- time grid size
- nonlinear residuals

Guillaume

**Reply by GiudGiud** [reply]
looks like you can play with detection of contact pairs to prevent contact from being detected with automatic_pairing_distance
https://mooseframework.inl.gov/syntax/Contact/index.html

**Reply by vtnick** [reply]
I am able to get a converged solution in MOOSE. 

From the link you provided, my understanding is that is for automatic pairing correct? I would need to not use the primary and secondary keys but instead provide the surfaces in the auto pairing and let MOOSE figure it out that way?

**Reply by GiudGiud** [reply]
I think so
@recuero will know on setting contact parameters

**Comment by recuero** [comment]
> Is there a parameter I am missing in the contact action block that controls the nodal distance when checking for contact?

If the bodies come into contact, and you expect them not to be, I wouldn't look into contact or contact parameters. My suggestion: Remove contact and analyze your setup and model parameters so that they match the expected thermal expansion behavior.

**Reply by vtnick** [reply]
I can give that another try on the higher heat input case. My initial thought on looking at the contact parameters is because with a lower heat input case (the rest of the input file remaining the same), the thermal expansion lines up with hand calcs.  In that version of the problem the expected gap is still on the E-4 range so my initial thought was that at E-5 I was running up a tolerance somewhere.

I will rerun both cases without contact.

**Reply by recuero** [reply]
>  at E-5 I was running up a tolerance somewhere

There is no default parameter or behavior that would cause that. That could _maybe_ happen if you are using contact constraints and your tolerances are very loose, but it's rare, that's why I'd double check the results without contact. Also, if you output the mesh on a visualization software, you should be able to see if there is contact.

**Reply by vtnick** [reply]
I am still running the cases without contact. The lower heat input case can solve quickly, but the higher heat input seems to be having issues finishing the solve requiring me to ramp up the BC. I am able to get to about 80% of the final heat input before the simulation slows down and requires really small timesteps to continue through. From the output so far, it appears that the expansion and temperatures are lining up well with my hand calcs.

**Reply by recuero** [reply]
If I understand well, when you run this simulation you last mentioned the bodies do not come into contact. But in the statement of the problem at the top you mention they do come into contact: Do you have an image that shows this?

**Reply by vtnick** [reply]
I have a lower heat input that does not come into contact and a higher that still shouldn't come into contact. This is based on hand calcs and running in Ansys. For my higher heat input the bodies do come into contact.

I have attached 3 images. The simulation at the beginning, the end and the contact pressure plot. All for the higher heat input case. I am also using a postprocessor to output the displacements on the OD of the inner cylinder and the ID of the outer cylinder. These relative displacements also show the gap is closed.

![cylinders_contact](https://github.com/idaholab/moose/assets/5813132/ce63690d-eaa1-487a-b93e-6f5c63dff9e6)
![cylinders_t0](https://github.com/idaholab/moose/assets/5813132/1f3d90fa-a790-4cc0-b28c-33501b1b4d08)
![cylinders_tend](https://github.com/idaholab/moose/assets/5813132/ce6adc22-3a95-40da-b4e6-250956f742ae)

**Reply by recuero** [reply]
You have negative contact pressure?

**Reply by vtnick** [reply]
Yes, that also does not make much sense to me. But since they weren't supposed to be in contact anyway, I figured I would start with that problem first. Unless you think they are related?

![contact_zoom](https://github.com/idaholab/moose/assets/5813132/ae867929-b90c-4ca0-a1ef-86c2f5bcb98d)

**Reply by recuero** [reply]
Contact doesn't "attract" surfaces, so I am still confused as to why you might be getting contact (zero gap) only when you include the contact action. It seems that your simulation without contact didn't finish; I'd first make sure that, with that single change, i.e. removing the contact action, you don't get the zero gap result.

You can attach your input file and I can take a quick look.

**Reply by vtnick** [reply]
Yeah, I will let you know when the non-contact run finishes. 

I have attached the input file associated with the images previously. The only change other than the contact vs non-contact between what is running now and this is that I am building the mesh within moose right now so I can more easily play with the mesh density rather than going back and forth with cubit.

[closed_gap_contact_hex20.zip](https://github.com/idaholab/moose/files/12689997/closed_gap_contact_hex20.zip)

**Reply by vtnick** [reply]
It does not appear that the thermal expansion case is going to finish....or at least in a reasonable amount of time. It went to about 85% of the required input before slashing the time steps/ramping on the heat input. At about 4 hours now and its sitting at 91% and taking E-8 time steps. The lower heat (an order of magnitude less) was less than 5 minutes.

As of right now in the simulation based on the displacement values, the bodies are not in contact which is an improvement since I believe when I ran the contact solution they were already in contact at this point. 

I have tried running a thermal only solution at the expected final gap number and it seems to line up well with hand calcs as well. So is there something wrong with my expansion inputs? I would imagine if they were really wrong the other solution wouldn't work at all, but maybe I need to adjust something due to the amount of expansion?

**Reply by recuero** [reply]
> I believe when I ran the contact solution they were already in contact at this point.

I am really curious about that --if you find out anything else on this, please post. On a slightly different note, you can try using 'PENALTY' instead of 'KINEMATICS' in your `Contact` block and see whether, if the bodies come into contact, the negative contact pressure values go away.

> So is there something wrong with my expansion inputs? I would imagine if they were really wrong the other solution wouldn't work at all, but maybe I need to adjust something due to the amount of expansion?

I don't know, but the fact that your simulation is struggling with (what it seems at first sight) simple physics indicates that your setup or solver settings can be improved. One thing I would try along those lines is to switch from `NEWTON` to `PJFNK` in your executioner block.

**Reply by vtnick** [reply]
I think I have narrowed it down to a main issue:

For some reason there is a difference in the thermal expansion results between the cubit mesh and moose mesh. There are more elements/nodes/dofs in the cubit mesh and both meshes use HEX27 elements. The cubit mesh has more expansion than the moose mesh.  The moose mesh gives "correct" results. Basically, I think that the cubit solution was expanding too much too early and resulting in contact sooner. It wasn't as noticeable at the lower heat input, but the moose solution is much closer to the hand calcs. As the high increases the differences do as well and becomes more evident. 

I will try changing the solver to PJFNK and see if that helps to get the thermal expansion to finish.

**Reply by recuero** [reply]
> There are more elements/nodes/dofs in the cubit mesh and both meshes use HEX27 elemen

I don't fully understand the difference between the Cubit mesh and MOOSE mesh in this case. But happy to know that MOOSE gives satisfactory results.

Regarding the negative contact pressure, HEX20 elements are known to cause that type of issues with node-face contact. But we typically obtain reasonable results with HEX27 and node-face.

**Reply by vtnick** [reply]
Switching the solver type seems to have helped the convergence issue at the end of the solution.

What is the best tool to compare the mesh generated by Moose and one generated by Cubit? I am running a comparison on the solution for the higher heat input case, but for the lower I ran the following:

1. Moose generated
2. Moose generated---output via --mesh-only, imported into cubit and then exported back to disk
3. Cubit generate mesh

1 & 2 provide the same solution (thermal and expansion)....with number 3 giving an almost 12% greater thermal expansion. Aside from commenting out unused parameters and changing to using a file mesh the input files are identical. 

I think overall the answer to my original question is that my model was generating too much thermal expansion and thus was coming into contact. This increased thermal expansion, for now at least, seems to be a result of using a cubit generated mesh.

**Reply by lindsayad** [reply]
@GiudGiud has created a [MeshDiagnosticsGenerator](https://mooseframework.inl.gov/source/meshgenerators/MeshDiagnosticsGenerator.html). Perhaps that could catch something in the CUBIT mesh

**Reply by vtnick** [reply]
Thanks...I think I found my error in the cubit mesh (typo in selecting nodes for a nodeset used in BCs). Resulted in the two models being slightly different. 

I think overall the issue has been resolved...still having some convergence issues considering the relatively simple problem setup, but I imagine that is down to solver settings.
